File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        // 종류 구하기
        int size = new HashSet<>(Arrays.asList(gems)).size();
        
        int[] answer = new int[2];
        int len = Integer.MAX_VALUE, start = 0;
        
        Map<String, Integer> map = new HashMap<>();
        
        for(int end = 0; end < gems.length; end++){
            map.put(gems[end], map.getOrDefault(gems[end], 0) + 1);
        while (map.get(gems[start]) > 1) {
                map.put(gems[start], map.get(gems[start]) - 1);
                start++;
            }
            if(map.size() == size && len > (end - start)){
                len = end - start;
                answer[0] = start+1;
                answer[1] = end+1;
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    # 먼저 보석 종류 구하기
    # 사전으로 담고 사전에 있는 가장 작은 갯수가 1일때 return을 해버리기
    gems_set = set(gems)
    gems_dict = {}
    length = 1e9
    start = 0
    end = 0
    answer = []
    for i in range(len(gems)):
        if not gems[i] in gems_dict:
            gems_dict[gems[i]] = 0
        gems_dict[gems[i]] += 1
        while gems_dict[gems[start]] > 1:
            gems_dict[gems[start]] -= 1
            start += 1
        if len(gems_dict) == len(gems_set) and length > i - start:
            length = i - start
            answer = [start + 1, i + 1]
    return answer